### PR TITLE
Create directory paths for COPY thereby ensuring correct perms

### LINKED ISSUE #1534: COPY creates folders with very restricted permissions
This issue was initially wrongfully reported under #1516 . The problem is that Buildah creates the folders in path for `COPY` using `mode=700` unlike Docker that uses `mode=755`. This is a problem for two reasons: running as non-root and Docker compatibility:
- If you're running as non-root _inside the container_, you now need to chase down all `COPY` statements and fix permissions with a `chmod`
- If you're migrating off Docker to Buildah, your images may get broken in unexpected ways

**Steps to reproduce the issue:**
1. Create an empty file: `touch distutils.cfg`
2. Create a `Dockerfile` that performs a `COPY` into a nested non-existent path
```Dockerfile
FROM docker.io/ubuntu
COPY distutils.cfg /usr/lib/python3.7/distutils/distutils.cfg
#                           ^- 'python3.7' and 'distutils' folders have to be created
#                               to copy the file
```
3. Build an image: `buildah bud -t proof-buildah .`
4. Create a container from it: `buildah from proof-buildah`
5. Check permissions: `buildah run proof-buildah-working-container ls -lah /usr/lib | grep python`

**Describe the results you received:**
`python3.7` folder has `mode=700`. See the full CLI output below:
```
$ buildah bud -t proof-buildah .
STEP 1: FROM docker.io/ubuntu
STEP 2: COPY distutils.cfg /usr/lib/python3.7/distutils/distutils.cfg
STEP 3: COMMIT proof-buildah
Getting image source signatures
Copying blob 762d8e1a6054 skipped: already exists
Copying blob e45cfbc98a50 skipped: already exists
Copying blob d60e01b37e74 skipped: already exists
Copying blob b57c79f4a9f3 skipped: already exists
Copying blob c7a123878960 done
Copying config 5ee072a37f done
Writing manifest to image destination
Storing signatures
--> 5ee072a37f58ff245eb40a410039bc989e93d03eba6e43aee0ce1cf910a12a7f
$ buildah from proof-buildah
proof-buildah-working-container
$ buildah run proof-buildah-working-container ls -lah /usr/lib | grep python
drwx------ 3 root root 4.0K Apr 18 21:05 python3.7
$ docker run --rm proof-buildah ls -lah /usr/lib | grep python
drwx------  3 root root 4.0K Apr 18 21:05 python3.7
```

**Describe the results you expected:**
`python3.7` folder has `mode=755`.  See the Docker CLI output below (for comparison):
```
$ docker build -t proof-docker .
Sending build context to Docker daemon   2.56kB
Step 1/2 : FROM docker.io/ubuntu
 ---> 94e814e2efa8
Step 2/2 : COPY distutils.cfg /usr/lib/python3.7/distutils/distutils.cfg
 ---> c1cebcb50780
Successfully built c1cebcb50780
Successfully tagged proof-docker:latest
$ docker run --rm proof-docker ls -lah /usr/lib | grep python
drwxr-xr-x  3 root root 4.0K Apr 18 21:04 python3.7
```

**Output of `rpm -q buildah` or `apt list buildah`:**

```
Listing... Done
buildah/xenial,now 1.8-1~dev~ubuntu16.04.2~ppa5 amd64 [installed]
```

**Output of `buildah version`:**

```
Version:         1.8-dev
Go Version:      go1.10.4
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:
Git Commit:
Built:           Sat Apr 13 16:42:38 2019
OS/Arch:         linux/amd64
```

**Output of `podman version` if reporting a `podman build` issue:**

```
N/A
```

**Output of `cat /etc/*release`:**

```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.1 LTS"
NAME="Ubuntu"
VERSION="16.04.1 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.1 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
UBUNTU_CODENAME=xenial
```

**Output of `uname -a`:**

```
Linux scw-35eb3f 4.5.7-generic #1 SMP Tue Jul 12 09:41:14 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# storage.conf is the configuration file for all tools
# that share the containers/storage libraries
# See man 5 containers-storage.conf for more information

# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "overlay"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary read-write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# AdditionalImageStores is used to pass paths to additional read-only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers (currently overlay, zfs, vfs, btrfs)
size = ""

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"
```

### PR DESCRIPTION
**Pull request by EricHripko**
# Overview
This PR changes the behaviour of `COPY` to create directories if they don't exist. Doing it in `Builder.Add` means that destination directories are created with correct permissions. Resolves #1534.

# Testing
9 new tests added:
- `buildah add` with absolute path
- `buildah add` with relative path
- `COPY` with absolute path
- `COPY` with relative path
- `ADD` with absolute path
- `ADD` with relative path
- Multi-stage `COPY` with absolute path
- Multi-stage `COPY` with relative path
- Copy to the root of the file system (edge case for the new code)

Output:
```
...
 ✓ buildah version test
 ✓ buildah version current in .spec file Version
 ✓ buildah version current in .spec file changelog

312 tests, 11 failures, 5 skipped
```
11 tests that are failing (as far as I can see) rely on spinning up an image registry at `localhost`. Therefore, I don't think they're caused by my changes.

# Signing
Signed-off-by: Eric Hripko <ehripko@bloomberg.net>

### CODE REVIEW

File: add.go [hunk 1]
```diff
@@ -112,6 +112,12 @@ func (b *Builder) Add(destination string, extract bool, options AddAndCopyOption
 	hostOwner := idtools.IDPair{UID: int(hostUID), GID: int(hostGID)}
 	dest := mountPoint
 	if destination != "" && filepath.IsAbs(destination) {
+		dir, _ := filepath.Split(destination)
```
Comment (rhatdan) [inline add.go]: Why not use filepath.Dir(destination)?

**Comment by rhatdan** [comment]
@nalind @TomSweeneyRedHat @giuseppe @vrothberg PTAL

File: add.go (hunk 1 above)
Comment (EricHripko) [reply add.go]: Great suggestion! Updated the code to use `Dir` instead.
Sorry for the delay :see_no_evil:

**Comment by rhatdan** [comment]
Can you squash your two commits together.

@nalind @TomSweeneyRedHat @giuseppe @vrothberg PTAL

File: add.go [hunk 2]
```diff
@@ -112,6 +112,12 @@ func (b *Builder) Add(destination string, extract bool, options AddAndCopyOption
 	hostOwner := idtools.IDPair{UID: int(hostUID), GID: int(hostGID)}
 	dest := mountPoint
 	if destination != "" && filepath.IsAbs(destination) {
+		dir := filepath.Dir(destination)
+		if dir != "." && dir != "/" {
+			if err = idtools.MkdirAllAndChownNew(filepath.Join(dest, dir), 0755, hostOwner); err != nil {
+				return errors.Wrapf(err, "error creating directory %q", filepath.Join(dest, dir))
+			}
+		}
```
Comment (TomSweeneyRedHat) [inline add.go]: if destination is a symlink does this work if the symlink target exists or again if the symlink target doesn't exist?

Comment (EricHripko) [reply add.go]: Will check :+1:

Comment (EricHripko) [reply add.go]: All current `bud.bats` tests pass, which includes `bud copy to symlink`, which tests copying to a [symlink that exists](https://github.com/containers/buildah/blob/master/tests/bud/dest-symlink/Dockerfile#L9). To be extra safe, I've verified that symlink is preserved:
```
STEP 7: RUN ls -ltr /etc
total 172
...
lrwxrwxrwx    1 root     root            16 Jun  8 15:24 hbase -> /usr/local/hbase
STEP 8: RUN ls -ltr /usr/local/hbase
total 4
-rw-r--r--    1 root     root           229 Jun  8 15:24 Dockerfile
```
This should cover the first case.

Comment (EricHripko) [reply add.go]: For dangling symlink as destination, there wasn't a test present. The behaviour of Docker is that symlink is 'fixed' (i.e., path is resolved and created for `COPY`):
```
...
Step 3/5 : RUN ls -ltr /
 ---> Running in 42c15f182748
lrwxrwxrwx   1 root root    8 Jun  8 15:39 src -> /symlink
Removing intermediate container 42c15f182748
 ---> 310bbc52eaf7
Step 4/5 : COPY file /src/
 ---> c07aca50f4ef
Step 5/5 : RUN ls -ltr / && ls /src
 ---> Running in aaae074b5b4b
total 68
...
lrwxrwxrwx   1 root root    8 Jun  8 15:39 src -> /symlink
drwxr-xr-x   2 root root 4096 Jun  8 15:39 symlink
...
file
Removing intermediate container aaae074b5b4b
 ---> 74191ba0e053
Successfully built 74191ba0e053
```

This is the same behaviour as `buildah` exhibits with this PR applied:
```
...
STEP 3: RUN ls -ltr /
total 64
...
lrwxrwxrwx   1 root root    8 Jun  8 15:35 src -> /symlink
...
STEP 4: COPY file /src/file
STEP 5: RUN ls -ltr / && ls /src
total 68
...
lrwxrwxrwx   1 root root    8 Jun  8 15:35 src -> /symlink
drwxr-xr-x   2 root root 4096 Jun  8 15:35 symlink
...
file
STEP 6: COMMIT
Getting image source signatures
Copying blob 762d8e1a6054 skipped: already exists
Copying blob e45cfbc98a50 skipped: already exists
Copying blob d60e01b37e74 skipped: already exists
Copying blob b57c79f4a9f3 skipped: already exists
Copying blob 2512810ddbad done
Copying config b156c22ae4 done
Writing manifest to image destination
Storing signatures
--> b156c22ae4bbce876c9daa5d90ef7f40ba1ef125d45af10fa03f0062e58baffa
```
Will add a test for this.

Comment (EricHripko) [reply add.go]:
```
✓ bud copy to dangling symlink
```
Done! This should address this comment :+1:

**Comment by EricHripko** [comment]
@rhatdan squashed the commits :+1:

**Comment by TomSweeneyRedHat** [comment]
LGTM, thx @EricHripko !

**Comment by rhatdan** [comment]
@rh-atomic-bot r+
